### ISSUE TITLE
Decrease the number of pages where we generate session on first load

### ISSUE THREAD

**Issue by ceckoslab** [nexcess/magento-turpentine#924]
I am working on a project that is about to get announced on many tv channels. We expect many concurrent users during the first seconds / minutes when there is a tv spot. We already tried with some performance testing and event there were some test tv spots in order to check how Magento performs.

We noticed performance issue because many sessions were created when the homepage was hit from customers who came from tv spot in same moment.

I took a look at the VCL file and got an idea that may reduce the number of generated sessions. If the customer has no frontend cookie to "pipe" only if Request Url match pages where a form_key / session is needed:
 * home page url
 *  ".html" (only product page urls end with ".html" and we need frontend cookie and form_key in order to add products in cart)
 * /customer/*
 * /contacts
 * /review/product/list/*
 * /checkout/*

I think that a condition can be put after "# it's a real user, make up a new session for them". I think in this condition we could check if Request Url matches any of the the urls listed above. If YES then  "pipe", if NO then serve page from cache
```
        if (req.http.Cookie !~ "frontend=" && !req.http.X-Varnish-Esi-Method) {
            if (client.ip ~ crawler_acl ||
                    req.http.User-Agent ~ "^(?:{{crawler_user_agent_regex}})$") {
                # it's a crawler, give it a fake cookie
                set req.http.Cookie = "frontend=crawler-session";
            } else {
                # it's a real user, make up a new session for them
                {{generate_session}}# call generate_session;
                return (pipe);
            }
        }
```

What do you think about this idea?

**Comment by miguelbalparda** [comment]
Have you tried using different session generation approaches? We currently have 3, see #846 for some more info.

**Comment by miguelbalparda** [comment]
Any updates on this @ceckoslab?

**Comment by ceckoslab** [comment]
I am working on solution that doesn't need the VCL fix. Also I have full control over the theme and I don't have add to cart actions by GET requests. I made a list of pages that can be considered as "public" and if customer doesn't have frontend cookie I do:

```
set req.http.X-Varnish-Faked-Session = "frontend=guest-customer-session";
```

Later, in next requests I check if I need real cookie for e.g. a customer reached product page, then I do:

```
# we need to create real session for the guest user
if (req.http.Cookie ~ "frontend=guest-customer-session" && req.url ~ "(\.html|/catalog/product/view)") {
            unset req.http.Cookie;
            unset req.http.X-Varnish-Faked-Session;
            return (pipe);
}
```

I also have to modify a bit Nexcessnet_Turpentine_Model_Varnish_Configurator_Abstract in order make this feature more configurable.

Basically my solution would work good in my context because I have full control. I may come with PR but I am not sure that it would work good in default Magento installation with default theme.

Do you have any thoughts?

**Comment by miguelbalparda** [comment]
Code it and we can see together the outcome. Not all PRs are suitable for everyone but probably someone is looking for the exact solution you are building.
